File: 912_Sort_an_Array/bubbleSort.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        int len = nums.size();
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len - i - 1; j++) {
                if (nums[j] > nums[j + 1]) {
                    int tmp = nums[j];
                    nums[j] = nums[j + 1];
                    nums[j + 1] = tmp;
                }
            }
        }
        return nums;
    }

    vector<int> sortArray2(vector<int>& nums) {
        int len = nums.size();
        for (int i = 0; i < len; i++) {
            bool sorted = true;
            for (int j = 0; j < len - i - 1; j++) {
                if (nums[j] > nums[j + 1]) {
                    sorted = false;
                    int tmp = nums[j];
                    nums[j] = nums[j + 1];
                    nums[j + 1] = tmp;
                }
            }
            if (sorted) break;
        }
        return nums;
    }
};

int main() {
    Solution s;
    vector<int> nums {5,2,3,1};
    for (int i : s.sortArray(nums)) {
        cout << i << " ";
    }
    cout << "\n";
}

File: 912_Sort_an_Array/bubbleSort.java
class Solution {
    public int[] sortArray(int[] nums) {
        bubbleSort2(nums, nums.length);
        return nums;
    }

    private void bubbleSort(int[] nums, int len) {
        for (int i = len - 1; i > 0; i--) {
            for (int j = 0; j < i; j++) {
                if (nums[j] > nums[j+1]) {
                    swap(nums, j, j+1);
                }
            }
        }
    }

    private void bubbleSort2(int[] nums, int len) {
        for (int i = len - 1; i > 0; i--) {
            Boolean sorted = true;
            for (int j = 0; j < i; j++) {
                if (nums[j] > nums[j+1]) {
                    sorted = false;
                    swap(nums, j, j+1);
                }
            }
            if (sorted) break;
        }
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

public class leetcode {
    public static void main(String[] args) {
        Solution s = new Solution();
        int[] nums = {5,2,3,1,-1,-34,45};
        s.sortArray(nums);
        for (int i : nums) 
            System.out.print(i + " ");
        System.out.println();
    }
}
